core/src/test/java/org/infinispan/statetransfer/ReadAfterLostDataTest.java

```diff
@@ -0,0 +1,301 @@
+import org.infinispan.context.InvocationContext;
+import org.infinispan.context.InvocationContextFactory;
+import org.infinispan.factories.ComponentRegistry;
+import org.infinispan.functional.impl.FunctionalMapImpl;
+import org.infinispan.functional.impl.ReadOnlyMapImpl;
+import org.infinispan.functional.impl.ReadWriteMapImpl;
+import org.infinispan.test.MultipleCacheManagersTest;
+import org.infinispan.test.TestingUtil;
+import org.infinispan.test.fwk.CleanupAfterMethod;
+import org.infinispan.test.fwk.InCacheMode;
+import org.infinispan.test.fwk.TransportFlags;
+import org.infinispan.topology.CacheTopology;
+import org.infinispan.topology.ClusterTopologyManager;
+import org.infinispan.util.BlockingClusterTopologyManager;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.Test;
+
+/**
+ * Tests operations executed when the actual owners are lost, with partition handling disabled (AP mode).
+ */
+@Test(groups = "functional", testName = "statetransfer.ReadAfterLostDataTest")
+@InCacheMode(CacheMode.DIST_SYNC)
+@CleanupAfterMethod
+public class ReadAfterLostDataTest extends MultipleCacheManagersTest {
+   private List<Runnable> cleanup = new ArrayList<>();
+
+   @Override
+   protected void createCacheManagers() throws Throwable {
+      ConfigurationBuilder cb = new ConfigurationBuilder();
+      cb.clustering().cacheMode(cacheMode)
+            .partitionHandling().enabled(false);
+      createClusteredCaches(4, cb, new TransportFlags().withFD(true).withMerge(true));
+   }
+
+   @AfterMethod
+   protected void cleanup() {
+      cleanup.forEach(Runnable::run);
+      cleanup.clear();
+   }
+
+   public void testGet() throws Exception {
+      test(ReadAfterLostDataTest::get, false, false);
+   }
+
+   public void testGetBeforeTopologyUpdate() throws Exception {
+      test(ReadAfterLostDataTest::get, false, true);
+   }
+
+   public void testGetAll() throws Exception {
+      test(ReadAfterLostDataTest::getAll, false, false);
+   }
+
+   public void testGetAllBeforeTopologyUpdate() throws Exception {
+      test(ReadAfterLostDataTest::getAll, false, true);
+   }
+
+   public void testPut() throws Exception {
+      test(ReadAfterLostDataTest::put, true, false);
+   }
+
+   public void testRemove() throws Exception {
+      test(ReadAfterLostDataTest::remove, true, false);
+   }
+
+   public void testReplace() throws Exception {
+      test(ReadAfterLostDataTest::replace, true, false);
+   }
+
+   //TODO: We don't test put/remove/replace before topology update as with interceptor these commands
+   // invoke RpcManager.sendTo that does not throw an exception when the target is not in view anymore.
+   // These commands rely on next topology (that is blocked) causing the OutdatedTopologyException.
+   // We'd need to execute them in parallel, wait until all of them use RpcManager.sendTo and then unblock
+   // the topology change. That's rather too much tied to the actual implementation.
```

Comment:
I'm not sure what the "after" tests actually check... if the new topology was already installed, then the leavers are no longer in the cluster, and the lost keys now have proper owners.